server/src/main/java/org/apache/druid/curator/announcement/NodeAnnouncer.java

```diff
@@ -0,0 +1,350 @@
+   * @param removeParentIfCreated remove parent of "path" if we had created that parent
+   */
+  public void announce(String path, byte[] bytes, boolean removeParentIfCreated)
+  {
+    synchronized (toAnnounce) {
+      if (!started) {
+        toAnnounce.add(new Announceable(path, bytes, removeParentIfCreated));
+        return;
+      }
+    }
+
+    final String parentPath = ZKPathsUtils.getParentPath(path);
+    boolean buildParentPath = false;
+
+    byte[] value = announcedPaths.get(path);
+
+    if (value == null) {
+      try {
+        if (curator.checkExists().forPath(parentPath) == null) {
+          buildParentPath = true;
+        }
+      }
+      catch (Exception e) {
+        log.debug(e, "Problem checking if the parent existed, ignoring.");
+      }
+
+      // Synchronize to make sure that I only create a listener once.
+      synchronized (toAnnounce) {
+        if (!listeners.containsKey(path)) {
+          final NodeCache cache = new NodeCache(curator, path, true);
+          cache.getListenable().addListener(
+              () -> {
+                ChildData currentData = cache.getCurrentData();
+                if (currentData == null) {
+                  final byte[] value1 = announcedPaths.get(path);
+                  if (value1 != null) {
+                    log.info("Node[%s] dropped, reinstating.", path);
+                    createAnnouncement(path, value1);
+                  }
+                }
+              }
+          );
+
+          if (started) {
+            if (buildParentPath) {
+              createPath(parentPath, removeParentIfCreated);
+            }
+            startCache(cache);
+            listeners.put(path, cache);
+          }
+        }
+      }
+    }
+
+    boolean created = false;
+    synchronized (toAnnounce) {
+      if (started) {
+        byte[] oldBytes = announcedPaths.putIfAbsent(path, bytes);
+
+        if (oldBytes == null) {
+          created = true;
+        } else if (!Arrays.equals(oldBytes, bytes)) {
+          throw new IAE("Cannot reannounce different values under the same path");
+        }
+      }
+    }
+
+    if (created) {
+      try {
+        createAnnouncement(path, bytes);
+      }
+      catch (Exception e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  public void update(final String path, final byte[] bytes)
+  {
+    synchronized (toAnnounce) {
+      if (!started) {
+        // removeParentsIfCreated is not relevant for updates; use dummy value "false".
+        toUpdate.add(new Announceable(path, bytes, false));
+        return;
+      }
+    }
+
+    byte[] oldBytes = announcedPaths.get(path);
+
+    if (oldBytes == null) {
+      throw new ISE("Cannot update a path[%s] that hasn't been announced!", path);
+    }
+
+    synchronized (toAnnounce) {
+      try {
+        if (!Arrays.equals(oldBytes, bytes)) {
+          announcedPaths.put(path, bytes);
+          updateAnnouncement(path, bytes);
```

Comment:
You can move this statement out of `synchronized ()` above to reduce locking time

Reply:
Do you mean that `synchronized` just needs to surround the `announcedPaths.put(path, bytes);`?

Reply:
It may be similar to `announce()` where you define a boolean variable, assign it within a critical section, and then use outside of a critical section to emit curator operation.

Reply:
I guess it would be probably worth to define the concurrency control more precisely across all variables and methods in this class. I'm not sure why `listeners` is a concurrentHashMap since it's guarded by `toAnnounce`. Similar question for `announcedPaths`. Seems like it's a concurrentHashMap to allow reads without locking `toAnnounce`. However, `unannounce()` updates `announcedPaths` without the lock on `toAnnounce`. Maybe they don't have to be concurrentHashMaps. Or maybe there is another way to provide a better concurrency control.

Reply:
@asdf2014 would you please explain and add a comment about why `listeners` and `announcedPaths` are concurrentHashMaps? And also why does `unannounce()` update `announcedPaths` without lock?